File: src/com/patterns/slidingwindow/medium/java/LongestSubStringWithoutRepeatingChars.java
package com.patterns.slidingwindow.medium.java;

import java.util.HashMap;

public class LongestSubStringWithoutRepeatingChars {

    public static void main(String[] args) {

        System.out.println(answer("abcabcbb"));
        System.out.println(answer("bbbbb"));
        System.out.println(answer("pwwkew"));
        System.out.println(answer("au"));
        System.out.println(answer(" "));

    }

    private static int answer(String s) {

        HashMap<Character, Integer> map = new HashMap<>();

        int windowStart = 0;
        int windowEnd = 0;
        int ans = 0;

        while (windowEnd < s.length()) {

            if (!map.containsKey(s.charAt(windowEnd))) {

                map.put(s.charAt(windowEnd), 0);
                windowEnd++;

            } else {

                ans = Math.max(ans, windowEnd - windowStart);
                map.clear();
                windowEnd = ++windowStart;

            }

        }

        return Math.max(ans, windowEnd - windowStart);

    }

}

File: src/com/patterns/slidingwindow/medium/python/longest_substring_without_repeating_chars.py
# LeetCode Q3
# Given a string s, find the length of the longest substring without repeating characters.

# Example 1:
# Input: s = "abcabcbb"
# Output: 3
# Explanation: The answer is "abc", with the length of 3.

# Example 2:
# Input: s = "bbbbb"
# Output: 1
# Explanation: The answer is "b", with the length of 1.

# Example 3:
# Input: s = "pwwkew"
# Output: 3
# Explanation: The answer is "wke", with the length of 3.


def answer(s: str) -> int:

    d = {}

    window_start = 0
    window_end = 0
    ans = 0

    while window_end < len(s):

        if s[window_end] not in d:

            d[s[window_end]] = 0
            window_end += 1

        else:

            ans = max(ans, window_end - window_start)
            d = {}
            window_start += 1
            window_end = window_start

    return max(ans, window_end - window_start)


print(answer("abcabcbb"))
print(answer("bbbbb"))
print(answer("pwwkew"))
print(answer("au"))
print(answer(" "))
